src/services/navigateTo.ts

```diff
@@ -1,35 +1,42 @@
 module ts.NavigateTo {
-    type RawNavigateToItem = { name: string; fileName: string; matchKind: MatchKind; declaration: Declaration };
-
-    enum MatchKind {
-        none = 0,
-        exact = 1,
-        substring = 2,
-        prefix = 3
-    }
-
-    export function getNavigateToItems(program: Program, cancellationToken: CancellationTokenObject, searchValue: string, maxResultCount: number): NavigateToItem[]{
-        // Split search value in terms array
-        var terms = searchValue.split(" ");
-
-        // default NavigateTo approach: if search term contains only lower-case chars - use case-insensitive search, otherwise switch to case-sensitive version
-        var searchTerms = map(terms, t => ({ caseSensitive: hasAnyUpperCaseCharacter(t), term: t }));
+    type RawNavigateToItem = { name: string; fileName: string; matchKind: PatternMatchKind; isCaseSensitive: boolean; declaration: Declaration };
 
+    export function getNavigateToItems(program: Program, cancellationToken: CancellationTokenObject, searchValue: string, maxResultCount: number): NavigateToItem[] {
+        var patternMatcher = createPatternMatcher(searchValue);
         var rawItems: RawNavigateToItem[] = [];
 
         // Search the declarations in all files and output matched NavigateToItem into array of NavigateToItem[] 
         forEach(program.getSourceFiles(), sourceFile => {
             cancellationToken.throwIfCancellationRequested();
 
-            var fileName = sourceFile.fileName;
             var declarations = sourceFile.getNamedDeclarations();
             for (var i = 0, n = declarations.length; i < n; i++) {
                 var declaration = declarations[i];
-                // TODO(jfreeman): Skip this declaration if it has a computed name
-                var name = (<Identifier>declaration.name).text;
-                var matchKind = getMatchKind(searchTerms, name);
-                if (matchKind !== MatchKind.none) {
-                    rawItems.push({ name, fileName, matchKind, declaration });
+                var name = getDeclarationName(declaration);
+                if (name !== undefined) {
+
+                    // First do a quick check to see if the name of the declaration matches the 
+                    // last portion of the (possibly) dotted name they're searching for.
+                    var matches = patternMatcher.getMatchesForLastSegmentOfPattern(name);

```

Comment:
As per discussion, the pattern matcher is concerned only with the domain of strings.  It can't know the candidate container unless it is given it.  And we don't want to compute the container if we can avoid it.